x-pack/plugins/actions/server/routes/get_oauth_access_token.ts

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { schema, TypeOf } from '@kbn/config-schema';
+import { IRouter, Logger } from '@kbn/core/server';
+import { ILicenseState } from '../lib';
+import { INTERNAL_BASE_ACTION_API_PATH } from '../../common';
+import { ActionsRequestHandlerContext } from '../types';
+import { verifyAccessAndContext } from './verify_access_and_context';
+import { ActionsConfigurationUtilities } from '../actions_config';
+import {
+  getOAuthJwtAccessToken,
+  GetOAuthJwtConfig,
+  GetOAuthJwtSecrets,
+} from '../builtin_action_types/lib/get_oauth_jwt_access_token';
+import {
+  getOAuthClientCredentialsAccessToken,
+  GetOAuthClientCredentialsConfig,
+  GetOAuthClientCredentialsSecrets,
+} from '../builtin_action_types/lib/get_oauth_client_credentials_access_token';
+
+const oauthJwtBodySchema = schema.object({
+  tokenUrl: schema.string(),
+  config: schema.object({
+    clientId: schema.string(),
+    jwtKeyId: schema.string(),
+    userIdentifierValue: schema.string(),
+  }),
+  secrets: schema.object({
+    clientSecret: schema.string(),
+    privateKey: schema.string(),
+    privateKeyPassword: schema.maybe(schema.string()),
+  }),
+});
+
+type OAuthJwtParams = TypeOf<typeof oauthJwtBodySchema>;
+
+const oauthClientCredentialsBodySchema = schema.object({
+  tokenUrl: schema.string(),
+  scope: schema.string(),
+  config: schema.object({
+    clientId: schema.string(),
+    tenantId: schema.string(),
+  }),
+  secrets: schema.object({
+    clientSecret: schema.string(),
+  }),
+});
+
+type OAuthClientCredentialsParams = TypeOf<typeof oauthClientCredentialsBodySchema>;
+
+const bodySchema = schema.object({
+  type: schema.oneOf([schema.literal('jwt'), schema.literal('client')]),
+  options: schema.conditional(
+    schema.siblingRef('type'),
+    schema.literal('jwt'),
+    oauthJwtBodySchema,
+    oauthClientCredentialsBodySchema
+  ),
+});
+
+export const getOAuthAccessToken = (
+  router: IRouter<ActionsRequestHandlerContext>,
+  licenseState: ILicenseState,
+  logger: Logger,
+  configurationUtilities: ActionsConfigurationUtilities
+) => {
+  router.post(
+    {
+      path: `${INTERNAL_BASE_ACTION_API_PATH}/connector/_oauth_access_token`,
+      validate: {
+        body: bodySchema,
+      },
+    },
+    router.handleLegacyErrors(
+      verifyAccessAndContext(licenseState, async function (context, req, res) {
+        const { type, options } = req.body;
+
```

Comment:
Should we limit access to this API to users who can edit connectors? I realize in the current state, non-Kibana users can access this endpoint and create tokens.

From the pattern in actions client (`await this.authorization.ensureAuthorized('update');`) should we also audit log this? cc @kobelb the API creates tokens detached from a connector, should we track an audit of these API calls somehow?

Reply:
> Should we limit access to this API to users who can edit connectors? I realize in the current state, non-Kibana users can access this endpoint and create tokens.

Good idea, I agree we should limit access in this way :+1:

> From the pattern in actions client (`await this.authorization.ensureAuthorized('update');`) should we also audit log this? cc @kobelb the API creates tokens detached from a connector, should we track an audit of these API calls somehow?

I'm a bit on the fence on this --
On one hand, it would be great to audit anything and everything that happens in Kibana 🎉 

On the other hand:
1. I question how useful a specific audit record for this would be. We already have audit records for HTTP requests, so we know user X made a request to this API endpoint. The only additional thing a new specific audit event would give an administrator is the ability to determine what was in the payload (`tokenUrl`, `scope`, etc.)
2. I'm not sure what "Category" such a new audit event would fit into. Our current audit event categories are "authentication", "database", and "web".

I propose that we rely on the existing HTTP request auditing for this event, and we just add a regular log message with a "Debug" log level that contains the non-secret config values in the request payload (`tokenUrl`, `scope`, etc) and the outcome (success or fail).

Reply:
@mikecote @jportner Restricted access to users with ability to edit connectors in this commit: https://github.com/elastic/kibana/pull/131171/commits/79962f5e4bbc1564ffb9e1fd40b2d94c42a7b46b

Also debug logging the success/failure of token generation.

Reply:
Thank you! Commit LGTM!